File: pkg/runtime/runtimes/ts/pulserpc/contract.ts
/**
 * Contract class for IDL validation and interface metadata
 *
 * This module provides the Contract class which parses IDL metadata
 * and provides validation for requests and responses.
 */

import { TypeDef, StructMap, EnumMap, findStruct, findEnum } from "./types.js";
import { validateType } from "./validation.js";

export interface FunctionDef {
  name: string;
  parameters: Array<{
    name: string;
    type: TypeDef;
    optional?: boolean;
  }>;
  returnType?: TypeDef;
  returnOptional?: boolean;
}

export interface Interface {
  name: string;
  functions: Map<string, FunctionDef>;
  getFunction(funcName: string): FunctionDef | undefined;
}

export interface JsonRpcIdlFormat {
  interfaces: Array<{
    name: string;
    methods: FunctionDef[];
  }>;
  structs: Array<{
    name: string;
    extends?: string;
    fields: Array<{
      name: string;
      type: TypeDef;
      optional?: boolean;
    }>;
  }>;
  enums: Array<{
    name: string;
    values: Array<{ name: string }>;
  }>;
}

/**
 * Represents an interface from the IDL
 */
class InterfaceImpl implements Interface {
  name: string;
  functions: Map<string, FunctionDef>;

  constructor(ifaceData: { name: string; methods: FunctionDef[] }) {
    this.name = ifaceData.name;
    this.functions = new Map();
    for (const func of ifaceData.methods || []) {
      this.functions.set(func.name, func);
    }
  }

  getFunction(funcName: string): FunctionDef | undefined {
    return this.functions.get(funcName);
  }
}

/**
 * Represents a parsed IDL contract
 *
 * The Contract class parses IDL JSON and provides validation
 * for requests and responses based on the interface definitions.
 */
export class Contract {
  idlParsed: any;
  interfaces: Map<string, Interface> = new Map();
  structs: StructMap = {};
  enums: EnumMap = {};

  constructor(idlParsed: any) {
    this.idlParsed = idlParsed;

    if (typeof idlParsed === 'object' && !Array.isArray(idlParsed)) {
      // PulseRPC format - dict with interfaces, structs, enums keys
      // Parse interfaces
      for (const ifaceData of idlParsed.interfaces || []) {
        this.interfaces.set(ifaceData.name, new InterfaceImpl(ifaceData));
      }

      // Parse structs
      for (const structData of idlParsed.structs || []) {
        this.structs[structData.name] = structData;
      }

      // Parse enums
      for (const enumData of idlParsed.enums || []) {
        this.enums[enumData.name] = enumData;
      }
    } else if (Array.isArray(idlParsed)) {
      // Barrister format - list of items with type field
      for (const item of idlParsed) {
        const itemType = item.type;
        if (itemType === 'struct') {
          this.structs[item.name] = item;
        } else if (itemType === 'enum') {
          this.enums[item.name] = item;
        } else if (itemType === 'interface') {
          this.interfaces.set(item.name, new InterfaceImpl(item));
        }
      }
    }
  }

  hasInterface(ifaceName: string): boolean {
    return this.interfaces.has(ifaceName);
  }

  getInterface(ifaceName: string): Interface | undefined {
    return this.interfaces.get(ifaceName);
  }

  /**
   * Validate request parameters against IDL
   */
  validateRequest(ifaceName: string, funcName: string, params: any[]): void {
    const iface = this.getInterface(ifaceName);
    if (!iface) {
      throw new Error(`Unknown interface: '${ifaceName}'`);
    }

    const func = iface.getFunction(funcName);
    if (!func) {
      throw new Error(`${ifaceName}: Unknown function: '${funcName}'`);
    }

    const paramDefs = func.parameters || [];

    // Check parameter count
    if (params.length !== paramDefs.length) {
      throw new Error(
        `Function '${ifaceName}.${funcName}' expects ${paramDefs.length} param(s). ${params.length} given.`
      );
    }

    // Validate each parameter
    for (let i = 0; i < params.length; i++) {
      const paramValue = params[i];
      const paramDef = paramDefs[i];
      const paramName = paramDef.name;
      const paramType = paramDef.type;
      const isOptional = paramDef.optional || false;

      try {
        validateType(paramValue, paramType, this.structs, this.enums, isOptional);
      } catch (e: any) {
        throw new Error(
          `Function '${ifaceName}.${funcName}' invalid param '${paramName}'. ${e.message}`
        );
      }
    }
  }

  /**
   * Validate response result against IDL
   */
  validateResponse(ifaceName: string, funcName: string, result: any): void {
    const iface = this.getInterface(ifaceName);
    if (!iface) {
      throw new Error(`Unknown interface: '${ifaceName}'`);
    }

    const func = iface.getFunction(funcName);
    if (!func) {
      throw new Error(`${ifaceName}: Unknown function: '${funcName}'`);
    }

    // Check if function has a return type
    const returnType = func.returnType;
    if (!returnType) {
      // Function returns void/None
      if (result !== null && result !== undefined) {
        throw new Error(
          `Function '${ifaceName}.${funcName}' invalid response: '${JSON.stringify(result)}'. Expected null/undefined`
        );
      }
      return;
    }

    // Validate return type
    const isOptional = func.returnOptional || false;
    try {
      validateType(result, returnType, this.structs, this.enums, isOptional);
    } catch (e: any) {
      throw new Error(
        `Function '${ifaceName}.${funcName}' invalid response: '${JSON.stringify(result)}'. ${e.message}`
      );
    }
  }
}

File: pkg/runtime/runtimes/python2/pulserpc/contract.py
"""Contract class for IDL validation and interface metadata

This module provides the Contract class which parses IDL metadata
and provides validation for requests and responses.
"""

from validation import validate_type


class Interface(object):
    """Represents an interface from the IDL"""

    def __init__(self, iface_data):
        """Initialize Interface from IDL data"""
        self.name = iface_data['name']
        self.functions = {}
        for func in iface_data.get('methods', []):
            self.functions[func['name']] = func

    def get_function(self, func_name):
        """Get function metadata by name"""
        return self.functions.get(func_name)


class Contract(object):
    """Represents a parsed IDL contract

    The Contract class parses IDL JSON and provides validation
    for requests and responses based on the interface definitions.
    """

    def __init__(self, idl_parsed):
        """Initialize Contract from parsed IDL"""
        if isinstance(idl_parsed, dict):
            self.idl_parsed = idl_parsed
            self.interfaces = {}
            self.structs = {}
            self.enums = {}
            self.meta = {}

            for iface_data in idl_parsed.get('interfaces', []):
                self.interfaces[iface_data['name']] = Interface(iface_data)

            for struct_data in idl_parsed.get('structs', []):
                self.structs[struct_data['name']] = struct_data

            for enum_data in idl_parsed.get('enums', []):
                self.enums[enum_data['name']] = enum_data
        else:
            self.idl_parsed = idl_parsed
            self.interfaces = {}
            self.structs = {}
            self.enums = {}
            self.meta = {}

            for item in idl_parsed:
                item_type = item.get('type')
                if item_type == 'struct':
                    self.structs[item['name']] = item
                elif item_type == 'enum':
                    self.enums[item['name']] = item
                elif item_type == 'interface':
                    self.interfaces[item['name']] = Interface(item)
                elif item_type == 'meta':
                    for key, value in item.items():
                        if key != 'type':
                            self.meta[key] = value

    def has_interface(self, iface_name):
        """Check if interface exists"""
        return iface_name in self.interfaces

    def get_interface(self, iface_name):
        """Get interface by name"""
        return self.interfaces.get(iface_name)

    def validate_request(self, iface_name, func_name, params):
        """Validate request parameters against IDL"""
        interface = self.get_interface(iface_name)
        if not interface:
            raise ValueError("Unknown interface: '%s'" % iface_name)

        func = interface.get_function(func_name)
        if not func:
            raise ValueError("%s: Unknown function: '%s'" % (iface_name, func_name))

        param_defs = func.get('parameters', [])

        if len(params) != len(param_defs):
            raise ValueError(
                "Function '%s.%s' expects %d param(s). %d given." %
                (iface_name, func_name, len(param_defs), len(params))
            )

        for i, param_value in enumerate(params):
            param_def = param_defs[i]
            param_name = param_def['name']
            param_type = param_def['type']
            is_optional = param_def.get('optional', False)

            try:
                validate_type(param_value, param_type, self.structs,
                            self.enums, is_optional)
            except (TypeError, ValueError) as e:
                raise ValueError(
                    "Function '%s.%s' invalid param '%s'. %s" %
                    (iface_name, func_name, param_name, e)
                )

    def validate_response(self, iface_name, func_name, result):
        """Validate response result against IDL"""
        interface = self.get_interface(iface_name)
        if not interface:
            raise ValueError("Unknown interface: '%s'" % iface_name)

        func = interface.get_function(func_name)
        if not func:
            raise ValueError("%s: Unknown function: '%s'" % (iface_name, func_name))

        return_type = func.get('returnType')
        if not return_type:
            if result is not None:
                raise ValueError(
                    "Function '%s.%s' invalid response: '%s'. Expected None" %
                    (iface_name, func_name, result)
                )
            return

        is_optional = func.get('returnOptional', False)
        try:
            validate_type(result, return_type, self.structs,
                        self.enums, is_optional)
        except (TypeError, ValueError) as e:
            raise ValueError(
                "Function '%s.%s' invalid response: '%s'. %s" %
                (iface_name, func_name, result, e)
            )
